util/EIP-2124/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! EIP-2124 implementation based on <https://eips.ethereum.org/EIPS/eip-2124>
+
+use crc::crc32;
+use ethereum_types::H256;
+use maplit::*;
+use rlp::{DecoderError, Rlp, RlpStream};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub type BlockNumber = u64;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ForkHash(pub u32);
+
+impl rlp::Encodable for ForkHash {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.encoder().encode_value(&self.0.to_be_bytes());
+	}
+}
+
+impl rlp::Decodable for ForkHash {
+	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+		rlp.decoder().decode_value(|b| {
+			if b.len() != 4 {
+				return Err(DecoderError::RlpInvalidLength);
+			}
+
+			let mut blob = [0; 4];
+			blob.copy_from_slice(&b[..]);
+
+			return Ok(Self(u32::from_be_bytes(blob)))
+		})
+	}
+}
+
+impl From<H256> for ForkHash {
+	fn from(genesis: H256) -> Self {
+		Self(crc32::checksum_ieee(&genesis[..]))
+	}
+}
+
+impl std::ops::AddAssign<BlockNumber> for ForkHash {
+	fn add_assign(&mut self, height: BlockNumber) {
+		let blob = height.to_be_bytes();
+		self.0 = crc32::update(self.0, &crc32::IEEE_TABLE, &blob)
+	}
+}
+
+impl std::ops::Add<BlockNumber> for ForkHash {
+	type Output = Self;
+	fn add(mut self, height: BlockNumber) -> Self {
+		self += height;
+		self
+	}
+}
+
+/// A fork identifier as defined by EIP-2124.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ForkId {
+	/// CRC32 checksum of the all fork blocks from genesis.
+	pub hash: ForkHash,     
+	/// Next upcoming fork block number, 0 if not yet known.
+	pub next: BlockNumber
+}
+
+impl rlp::Encodable for ForkId {
```

Comment:
this can be derived with rlp_derive https://github.com/paritytech/parity-ethereum/blob/06138ec7d3309c6f1a1230b2118e63aeb7e138c7/util/rlp-derive/tests/rlp.rs#L20

Reply:
done